File: backend/controllers/leaderboardController.js
const redisClient = require('../config/client');
const { getLatestLeaderboard } = require('../utils/redisHelper');

// Fetch the leaderboard
const getLeaderboard = async (req, res) => {
  try {
    const leaderboard = await getLatestLeaderboard();
    res.status(200).json(leaderboard);
  } catch (error) {
    console.error('Error fetching leaderboard:', error);
    res.status(500).json({ message: 'Internal Server Error' });
  }
};

// Clear the leaderboard (optional functionality)
const clearLeaderboard = async (req, res) => {
  try {
    await redisClient.del('leaderboard');
    req.io.emit('leaderboardUpdate', []); // Emit empty leaderboard after clearing
    res.status(200).json({ message: 'Leaderboard cleared successfully' });
  } catch (error) {
    console.error('Error clearing leaderboard:', error);
    res.status(500).json({ message: 'Internal Server Error' });
  }
};

module.exports = {
  getLeaderboard,
  clearLeaderboard
};

File: backend-golang/controllers/leaderboard_controller.go
package controllers

import (
	"encoding/json"
	"net/http"

	"backend-golang/utils"

	"github.com/go-redis/redis/v8"
	"golang.org/x/net/context"
)

var ctx = context.Background()

// Get leaderboard
func GetLeaderboard(redisClient *redis.Client) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		leaderboard, err := utils.GetLatestLeaderboard(redisClient)
		if err != nil {
			http.Error(w, "Error fetching leaderboard", http.StatusInternalServerError)
			return
		}
		json.NewEncoder(w).Encode(leaderboard)
	}
}

// Clear leaderboard
func ClearLeaderboard(redisClient *redis.Client) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		err := redisClient.Del(ctx, "leaderboard").Err()
		if err != nil {
			http.Error(w, "Error clearing leaderboard", http.StatusInternalServerError)
			return
		}
		w.WriteHeader(http.StatusOK)
	}
}
